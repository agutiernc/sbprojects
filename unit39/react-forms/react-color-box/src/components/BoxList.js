import React, { useState } from 'react';
import Box from './Box'
import NewBoxForm from './NewBoxForm';

const BoxList = () => {
  const [boxes, setBoxes] = useState([])

  const newBox = (boxObj) => setBoxes(boxes => [...boxes, boxObj])
  const removeBox = (id) => setBoxes(boxes => boxes.filter(b => b.id !== id))

  return (
    <div>
      <NewBoxForm newBox={newBox} />

      {
        boxes.map(b => (
          <Box
            key={b.id}
            id={b.id}
            width={b.width}
            height={b.height}
            bgColor={b.bgColor}
            handleRemove={removeBox}
          />
        ))
      }
    </div>
  )
}


export default BoxList;