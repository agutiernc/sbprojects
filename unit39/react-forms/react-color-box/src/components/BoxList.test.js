import React from "react";
import { render, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom'
import BoxList from './BoxList'

const newBox = (boxList, width='100', height='100', bgColor='red') => {
  const widthInput = boxList.getByLabelText('Width:')
  const heightInput = boxList.getByLabelText('Height:')
  const bgColorInput = boxList.getByLabelText('Background Color:')

  fireEvent.change(widthInput, { target: { value: width }})
  fireEvent.change(heightInput, { target: { value: height }})
  fireEvent.change(bgColorInput, { target: { value: bgColor }})

  const button = boxList.getByText('create new box')

  fireEvent.click(button)
}

// smoke test
it('renders without crashing', () => {
  render(<BoxList />)
})

// snapshot test
it('matches snapshot', () => {
  const { asFragment } = render(<BoxList />)

  expect(asFragment()).toMatchSnapshot()
})

it('can create a new box', () => {
  const boxList = render(<BoxList />)

  // with no boxes
  expect(boxList.queryByText('Remove Box')).not.toBeInTheDocument();
  
  newBox(boxList)

  const removeBtn = boxList.getByText('Remove Box')

  expect(removeBtn).toBeInTheDocument()
  expect(document.querySelector('#box')).toBeInTheDocument()

  expect(document.querySelector('#box')).toHaveStyle('background-color: red')
  expect(document.querySelector('#box')).toHaveStyle('width: 100px')
  expect(document.querySelector('#box')).toHaveStyle('height: 100px')
})

it('can remove box', () => {
  const boxList = render(<BoxList />)

  newBox(boxList)

  const removeBtn = boxList.getByText('Remove Box')

  fireEvent.click(removeBtn)

  expect(removeBtn).not.toBeInTheDocument()
})