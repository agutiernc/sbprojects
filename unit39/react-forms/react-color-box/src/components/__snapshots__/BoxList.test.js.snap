// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <div>
          <label
            for="width"
          >
            Width:
          </label>
          <input
            id="width"
            name="width"
            placeholder="100"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            for="height"
          >
            Height:
          </label>
          <input
            id="height"
            name="height"
            placeholder="100"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            for="bgColor"
          >
            Background Color:
          </label>
          <input
            id="bgColor"
            name="bgColor"
            placeholder="red"
            type="text"
            value=""
          />
        </div>
        <button>
          create new box
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
